chemprop/data/splitting.py

```diff
@@ -218,35 +217,45 @@ def _unpack_astartes_result(
     return list(train_idxs), list(val_idxs), list(test_idxs)
 
 
-def split_component(
-    datapointss: Sequence[Sequence[MoleculeDatapoint | MulticomponentDatapoint]],
-    split: SplitType | str = "random",
-    key_index: int = 0,
-    **kwargs,
+def split_data_by_indices(
+    data: Datapoints | MulticomponentDatapoints,
+    train_indices: Sequence[Sequence[int]] | Sequence[int] | None = None,
+    val_indices: Sequence[Sequence[int]] | Sequence[int] | None = None,
+    test_indices: Sequence[Sequence[int]] | Sequence[int] | None = None,
 ):
-    """Splits multicomponent data into training, validation, and test splits."""
-
-    key_datapoints = datapointss[key_index]
-    train_idxs, val_idxs, test_idxs = split_data(key_datapoints, split=split, **kwargs)
-
-    match SplitType.get(split):
-        case SplitType.CV_NO_VAL | SplitType.CV:
-            # convert indices to datapoints for each fold
-            train = [
-                [[datapoints[i] for i in fold] for datapoints in datapointss] for fold in train_idxs
-            ]
-            val = [
-                [[datapoints[i] for i in fold] for datapoints in datapointss] for fold in val_idxs
-            ]
-            test = [
-                [[datapoints[i] for i in fold] for datapoints in datapointss] for fold in test_idxs
+    """Splits data into training, validation, and test groups based on split indices given."""
+
+    train_data = _splitter_helper(data, train_indices) if train_indices is not None else None
+    val_data = _splitter_helper(data, val_indices) if val_indices is not None else None
+    test_data = _splitter_helper(data, test_indices) if test_indices is not None else None
+
+    return train_data, val_data, test_data
+
+
+def _splitter_helper(data, indices):
```

Comment:
`split_data_by_indices()` does dynamic splitting by calling `_splitter_helper()` which checks if the data and indices are lists or lists of lists. We don't love type checks, but I think it is safe to check `isinstance(data[0], (MoleculeDatapoint, ReactionDatapoint))` because we control the `Datapoint` types. More complicated was checking if the indices is a list of lists of ints or a lists of ints. `isinstance(indices[0], int)` doesn't work because `numpy.int64` is not a subtype of int. I settled on checking if `indices[0]` can be used as an index with a try, except. Is there a better way check if the lists are nested? I could also check if `data[0]` and `indices[0]` are lists or tuples, but that isn't required as they just need to be a `typing.Sequence`. We could also have four different splitting functions for the four cases and make the user choose, but that also doesn't seem right.

Reply:
```
>>> from collections.abc import Sequence
>>> a = [1, 2, 3]
>>> b = [[1, 2, 3], [4, 5, 6]]
>>> isinstance(a[0], Sequence)
False
>>> isinstance(b[0], Sequence)
True
```
Instead of checking for specific types , you can check for more generalized types using collections.abc which provides abstract base classes that can be used to test whether an object supports a particular interface.

Reply:
Unfortunately numpy arrays (which are used astartes) [don't](https://stackoverflow.com/questions/76082807/why-numpy-array-is-not-of-sequence-type) inherit from Sequence, so this simple check doesn't work. I'm still open to other ideas though because the current version still feels clunky to me.

Reply:
Can you try `Iterable` instead?

Reply:
```
>>> import numpy as np
>>> from collections.abc import Iterable
>>> a = np.array([1, 2, 3])
>>> b = np.array([[1, 2, 3], [4, 5, 6]])
>>> isinstance(a[0], Iterable)
False
>>> isinstance(b[0], Iterable)
True
```
I think `Iterable` would work in this case.